pymeasure/instruments/hp/hp3437A.py

```diff
@@ -0,0 +1,412 @@
+            trigger_mode = "external"
+        if trig == 3:
+            trigger_mode = "hold/manual"
+        return trigger_mode
+
+    @staticmethod
+    def _unpack_data(self, data):
+        """
+        Method to unpack the data from the returned bytes in packed mode
+
+        :param data: list of bytes to be decoded
+        :return ret_data: float value
+
+        """
+        ret_data = self.packed_data(self.packed_bytes(*data))
+        return float(ret_data.b)
+
+    # commands overwriting the base implementaiton
+    def read(self):
+        """
+        Reads measured data from instrument, returns a np.array.
+
+        (This function also takes care of unpacking the data if required)
+
+        :return data: np.array containing the data
+        """
+        # Adjusting the timeout to match the number of counts and the delay
+
+        current_timeout = self.adapter.connection.timeout
+        time_needed = self.number_readings * self.delay
+        new_timeout = time_needed * 3 * 1000  # safety factor 3
+        # TODO Review required! (2021/12/29)
+        if new_timeout > current_timeout:
+            if new_timeout >= 1e6:
+                # Disables timeout if measurement would take more then 1000 sec
+                new_timeout = 1000000
+                log.info("HP3437A: timeout deactivated")
+            self.adapter.connection.timeout = new_timeout
+            log.info("HP3437A: timeout changed to %g", new_timeout)
+        read_data = self.adapter.connection.read_raw()
+        # check if data is in packed format format
+        if self.talk_ascii is not True:
+            processed_data = []
+            read_data_length = int(len(read_data) / 2)
+            for i in range(0, read_data_length):
+                _from = i * 2
+                _to = _from + 2
+                processed_data.append(self._unpack_data(self, read_data[_from:_to]))
+            self.adapter.connection.timeout = current_timeout
+            return np.array(processed_data)
+        self.adapter.connection.timeout = current_timeout
+        return np.array(read_data[:-2].decode("ASCII").split(","), dtype=float)
+
+    # commands/properties for instrument control
+    def check_errors(self):
+        """
+        As this instrument does not have a error indication bit,
+        this function alwyas returns 0.
+
+        """
+        return 0
+
+    @property
+    def talk_ascii(self):
+        """Return True if the instrument is set to ASCII communciation,
+        this property can be set.
+
+        .. Note::
+
+            ASCII communciation is slower then the packed (BCD) communication,
+            this may cause problems with measurments when short delays are used
+
+
+
+        """
+        return bool(self._decode_status(self, self._fetch_status(), "Format"))
+
+    @talk_ascii.setter
+    def talk_ascii(self, value):
+        if value is True:
+            self.write("F1")
+        else:
+            self.write("F2")
+
```

Comment:
```
talk_ascii = Instrument.control("B", "%s",
                                   """ Return True if the instrument is set to ASCII communication,
        this property can be set.
                                   This property can be read
                                   """,
                                   values={True: "F1", False: "F2"},
                                   map_values = True,
                                   get_process=<process output of B and return "F1" or "F2">
                                   )
```
I was thinking about the above sketch implementation, but I believe there are reasons why this cannot be done.
Could you please comment on this point ?
Similar consideration are valid also for other properties.

Reply:
The problem on this device is that the status information queried with the `B` command is encoded in a raw byte string of 7 bytes. and I think from earlier tests instrument.control is not 8bit transparent as a `read_raw()` is.

I could try switch the `@xxx.setter` methods  to `Instrument.setting,` this would clean up the code a bit (?).

I will try a bit more maybe I come up with some good compromise solution.

Reply:
What about redefining `read `as a wrapper to `read_raw`, something like:
```
def read(self):
     return self.adapter.connection.read_raw()
```
In this mode, you could have ` get_process` parameter that can make the necessary post processing to extract information from binary value.
I think you need also to redefine the `values` though.

Reply:
@msmttchr that it is a good idea, I might try that later this week,